hera_sim/adjustment.py

```diff
@@ -0,0 +1,974 @@
+        values for the individual components, in ENU-coordinates.
+    relabel_antennas: bool, optional
+        Whether to rewrite antenna numbers/names when modifying the target
+        antenna array to optimally match the reference antenna array. Default
+        is to relabel the target object's antennas with the reference antenna
+        labels.
+    conjugation_convention: str, optional
+        Which conjugation convention to enforce after performing the antenna
+        array adjustment step. Default is to not enforce a convention.
+    overwrite_telescope_metadata: bool, optional
+        Whether to overwrite the target object's telescope metadata with that
+        of the reference data. Default is to leave the target object's telescope
+        metadata unchanged.
+    verbose: bool, optional
+        Whether to print brief statements noting progress. Default is to not
+        print updates.
+
+    Returns
+    -------
+    modified_data : :class:`pyuvdata.UVData` or :class:`Simulator` object
+        :class:`pyuvdata.UVData` object containing the modified target data.
+        If the initial target data was passed as a :class:`Simulator` object,
+        then the returned object is also a :class:`Simulator` object.
+
+    Notes
+    -----
+    The modified data will have its time and LST metadata modified to exactly
+    match the time and LST metadata of the reference data for all reference LSTs
+    that lie within the bounds of the target object's LSTs. Additionally, only a
+    subset of the original antenna array will remain, and the remaining antennas
+    will have their positions modified (either shifted or completely overwritten)
+    such that the modified positions give the largest intersection with the
+    reference antenna array, up to the specified tolerance. The flag and nsamples
+    arrays do not have their values modified, but may be reshaped.
+
+    Unless the integration times for the target and reference match each other
+    exactly, it is strongly recommended that you use cubic spline interpolation
+    rather than the rephasing provided here (which is just a thin wrapper around
+    a rephasing tool from :package:`hera_cal`), since small discrepancies in
+    integration times can result in some integrations effectively being skipped
+    by the rephasing tool, producing discontinuities in the rephased data.
+
+    Finally, a word of caution: if the target object has its data compressed by
+    redundancy, then it is possible that the antenna array adjustment step may
+    trigger a memory overflow. The current algorithm requires that every antenna
+    in the target object has data associated with it, and ensures that this is
+    true by inflating the data by redundancy. (It is not clear if it is even
+    possible to perform the antenna array adjustment step for an object with
+    arbitrary compression by redundancy.)
+    """
+    if verbose:
```

Comment:
I'll look into that! Thanks for the suggestion.

Reply:
So, on the `logging` issue... I keep banging my head on the wall trying to get this to work the way I would expect it to, while also being able to test for expected output, when using `logging` instead of these simplistic `print` statements. My first pass was to just replace the print statements with `logging.info` statements, and if the verbose setting was desired, then I exposed these to the console by setting `logging.basicConfig(format="%(message)s", level=logging.INFO)`. This worked, but it had the downside that every call to `adjust_to_reference` afterward, even with the verbose setting *off*, would have the logs printed. So I added the clause `logging.disable(logging.INFO)` that was executed if the verbose setting wasn't being used... but that caused logs to not be print in subsequent calls with the verbose setting on.

I suppose what I could do is retrieve the root logger, disable it if verbosity is disabled, then make sure it's enabled at the end of the function call... though that feels a little dangerous?

Reply:
logging can be annoying -- at least, to me (I'm not that familiar with it) -- but I think once you get used to the flow of it, it probably makes a lot more sense.

In this instance, I'd rely on the _user_ modifying the logger to print or not print the statements, rather than having a `verbose` keyword. If the logger is appropriately named (i.e. `logger = logging.getLogger(__name__)`) then the user should be able to turn things on and off themselves (either at root level or the level for this logger).

Reply:
Yeah, I was rather frustrated trying to learn how to work with it, though the code looks cleaner when I use `logging`. I guess the approach I described would be more sensible for the CLI. I'll see what I can do.

Reply:
I've done as you suggested: the `verbose` option is no longer available, and this function writes information to the logger before each step. The logging information is rather sparse, but the intention is folks who are using it in an interactive environment may just want short updates to print to inform them of how far the adjustment has progressed. It may be a good idea to raise an issue to add better logging support to help with benchmarking stuff (akin to the messages you added when we were hammering out some of the bugs in the end-to-end validation fileprep pipeline).